### PR TITLE
More tests and conn handling improvements

### PR DESCRIPTION
**Pull request by drawks**
- Added a bunch more tests to carbonlink
- Removed tests for logging side effects in favor of direct isolated unit
  tests #83 #87 
- modified carbonlink to immediately drop connection on malformed queries
- Enabled race detector in Makefile's test target

### CODE REVIEW

**Comment by drawks** [comment]
This PR should improve test coverage and reduce dependence on testing for exact language in errors and/or log messages.
The only behavioral change is that connections are dropped immediately on carbonlink queries that timeout while reading from the conn or fail on parsing the query packet. Both cases seemed like a sane behavior as there is no way to correct for receiving phase after an incomplete read.

**Comment by lomik** [comment]
Please, don't enable "-race" by default before https://github.com/Sirupsen/logrus/issues/295 fixed.

**Comment by drawks** [comment]
Tests should pass with race detection enabled, this PR handles the bug in logrus by removing calls to change the state of the logger so that bug doesn't trigger race detection.

File: cache/carbonlink.go [hunk 1]
```diff
@@ -133,13 +133,15 @@ func (listener *CarbonlinkListener) handleConnection(conn net.Conn) {
 
 		reqData, err := ReadCarbonlinkRequest(reader)
 		if err != nil {
+			conn.(*net.TCPConn).SetLinger(0)
```
Comment (redbaron) [inline cache/carbonlink.go:136]: shouldn't it call `conn.Close()` ? if message is malformed there are no sync markers in protocol to find beginning of new message, it all going to be garbage

**Comment by lomik** [comment]
> Tests should pass with race detection enabled, this PR handles the bug in logrus by removing calls to change the state of the logger so that bug doesn't trigger race detection.

Ok

There is another problem. Race detector greatly slows down tests. On my MacBook started to fall another test

```
--- FAIL: TestThrottleChan (8.01s)
        Location:       throttle_test.go:38
    Error:      Should be true
    Messages:   perSecond: 531234, bw: 347932

FAIL
```

**Comment by lomik** [comment]
In travis same problem

```
--- FAIL: TestThrottleChan (7.00s)
    Location:   throttle_test.go:38
    Error:      Should be true
    Messages:   perSecond: 213000, bw: 200221

FAIL
```

File: cache/carbonlink.go (hunk 1 above)
Comment (drawks) [reply cache/carbonlink.go:136]: The break after the message jumps out of the for loop which causes the rest of the function to run and then return at which point the deferred Close runs. If I've got the flow wrong here please advise.

**Comment by drawks** [comment]
The throttle chan test should really run as two steps, one where it runs unthrottled to determine the throughput in the given test env followed by a throttled run which sets the throttle to some significantly reduced rate(20% of the max or similar).

File: cache/carbonlink.go (hunk 1 above)
Comment (redbaron) [reply cache/carbonlink.go:136]: true, missed it

**Pull request merged**